fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/SpreadableBlockRegistryImpl.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.content.registry;
+
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
+import net.minecraft.registry.tag.TagKey;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.registry.SpreadableBlockRegistry;
+
+public final class SpreadableBlockRegistryImpl implements SpreadableBlockRegistry {
+	private static final Logger LOGGER = LoggerFactory.getLogger(SpreadableBlockRegistryImpl.class);
+
+	private static final Map<Block, Identifier> SPREADABLE_TYPE;
+	private static final Map<Identifier, SpreadableBlockRegistry> REGISTRIES;
+
+	static {
+		SPREADABLE_TYPE = new IdentityHashMap<>();
+		REGISTRIES = new HashMap<>();
+
+		SPREADABLE_TYPE.put(Blocks.GRASS_BLOCK, SpreadableBlockRegistry.GRASS);
+		SPREADABLE_TYPE.put(Blocks.MYCELIUM, SpreadableBlockRegistry.MYCELIUM);
+
+		REGISTRIES.put(SpreadableBlockRegistry.GRASS, new SpreadableBlockRegistryImpl(SpreadableBlockRegistry.GRASS));
+		REGISTRIES.put(SpreadableBlockRegistry.MYCELIUM, new SpreadableBlockRegistryImpl(SpreadableBlockRegistry.MYCELIUM));
+	}
+
+	public static SpreadableBlockRegistry getOrCreateInstance(Identifier type) {
+		return REGISTRIES.computeIfAbsent(type, SpreadableBlockRegistryImpl::new);
+	}
+
+	public static @Nullable SpreadableBlockRegistry getInstanceBySpreadable(@Nullable BlockState blockState) {
+		if (blockState == null) {
+			return null;
+		}
+
+		return getInstanceBySpreadable(blockState.getBlock());
+	}
+
+	public static @Nullable SpreadableBlockRegistry getInstanceBySpreadable(@Nullable Block block) {
+		if (block == null || !SPREADABLE_TYPE.containsKey(block)) {
```

Comment:
Why would block ever be null?

Reply:
I'm a C programmer.  I have trust issues.  :P
I've changed the methods to require non-null, in 11a0801.